shared-bindings/ota/__init__.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/ota/__init__.h"
+
+//| """OTA Module
+//|
+//| The `ota` module implements over-the-air update.
+//|
+//| There are two identical ota partitions ota_0/1, these
+//| contain different firmware versions.
+//| Having two partitions enables rollback functionality.
+//|
+//| The two partitions are defined as boot partition and
+//| next-update partition. Calling `ota.flash()` writes the
+//| next-update partition.
+//|
+//| After the next-update partition is written a validation
+//| check is performed and on a successful validation this
+//| partition is set as the boot partition. On next reset,
+//| firmware will be loaded from this partition"""
+//|
+
+//| def switch() -> None:
+//|     """Switches the boot partition."""
+//|     ...
+//|
+STATIC mp_obj_t ota_switch(void) {
+    common_hal_ota_switch();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(ota_switch_obj, ota_switch);
+
+//| def finish() -> None:
+//|     """Validates flashed firmware, sets next boot partition.
+//|         **Must be called after** `ota.flash()`"""
```

Comment:
Should this return a bool to indicate success or failure?

Reply:
Raising an error at the moment.

Reply:
Why? Errors are harder to document than return values. Are there multiple values you could raise?

Reply:
Yes, there are multiple errors that can be raised.

https://github.com/adafruit/circuitpython/blob/37ee5e683deb26ee691090d4f51b0ade4fa9b61c/ports/esp32s2/common-hal/ota/__init__.c#L127-L140

Reply:
I think this should be `switch_boot_bank()` and it should raise an exception if the firmware image in the other bank is invalid or the set fails. Make sure to raise a different exception for these two cases. I'm not sure the current code does. It appears to only log different messages.